File: simulation_py/node.py
"""
Node Simulation Module
Represents individual participants in the network
"""

import random
from effectiveness import calculate_ramp, calculate_decay


class Node:
    """Represents a single node in the network"""
    
    def __init__(self, node_id: str, join_day: int = 0, node_type: str = 'honest'):
        self.id = node_id
        self.join_day = join_day
        self.type = node_type
        
        # State tracking
        self.is_online = True
        self.participation_days = 0
        self.offline_days = 0
        self.effectiveness_when_offline = 0
        self.total_rewards = 0
        
        # History for analysis
        self.history = []
        
        # For intermittent nodes
        self.uptime_percent = 100
    
    def tick(self, current_day: int) -> float:
        """Update node state for a new day"""
        if current_day < self.join_day:
            return 0
        
        if self.is_online:
            self.participation_days += 1
            self.offline_days = 0
            effectiveness = calculate_ramp(self.participation_days)
        else:
            self.offline_days += 1
            effectiveness = calculate_decay(self.effectiveness_when_offline, self.offline_days)
            
            # If effectively reset, clear participation
            if effectiveness == 0:
                self.participation_days = 0
        
        self.history.append({
            "day": current_day,
            "effectiveness": effectiveness,
            "is_online": self.is_online,
            "participation_days": self.participation_days,
        })
        
        return effectiveness
    
    def go_offline(self):
        """Take node offline"""
        if self.is_online:
            self.effectiveness_when_offline = calculate_ramp(self.participation_days)
            self.is_online = False
    
    def go_online(self):
        """Bring node back online"""
        if not self.is_online:
            import math
            current_eff = calculate_decay(self.effectiveness_when_offline, self.offline_days)
            
            # Convert current effectiveness back to equivalent participation days
            if current_eff > 0:
                R = 40  # From config
                self.participation_days = -R * math.log(1 - current_eff)
            else:
                self.participation_days = 0
            
            self.is_online = True
            self.offline_days = 0
    
    def add_reward(self, amount: float):
        """Add rewards to node"""
        self.total_rewards += amount
    
    def get_effectiveness(self) -> float:
        """Get current effectiveness"""
        if not self.history:
            return 0
        return self.history[-1]["effectiveness"]
    
    def get_summary(self) -> dict:
        """Get node summary"""
        return {
            "id": self.id,
            "type": self.type,
            "join_day": self.join_day,
            "participation_days": self.participation_days,
            "is_online": self.is_online,
            "current_effectiveness": self.get_effectiveness(),
            "total_rewards": self.total_rewards,
        }


class NodeFactory:
    """Factory for creating different node types"""
    
    _counter = 0
    
    @classmethod
    def create_honest(cls, join_day: int = 0) -> Node:
        """Create an honest node that stays online"""
        cls._counter += 1
        return Node(f"honest-{cls._counter}", join_day, 'honest')
    
    @classmethod
    def create_intermittent(cls, join_day: int = 0, uptime_percent: int = 80) -> Node:
        """Create an intermittent node with random availability"""
        cls._counter += 1
        node = Node(f"intermittent-{cls._counter}", join_day, 'intermittent')
        node.uptime_percent = uptime_percent
        return node
    
    @classmethod
    def create_attacker(cls, join_day: int = 0) -> Node:
        """Create an attacker node"""
        cls._counter += 1
        return Node(f"attacker-{cls._counter}", join_day, 'attacker')
    
    @classmethod
    def reset(cls):
        """Reset counter"""
        cls._counter = 0

File: simulation/core/node.js
/**
 * Node Simulation Module
 * Represents individual participants in the network
 */

import { calculateRamp, calculateDecay } from './effectiveness.js';

/**
 * Represents a single node in the network
 */
export class Node {
    /**
     * Create a new node
     * @param {string} id - Unique node identifier
     * @param {number} joinDay - Day when node joined the network
     * @param {string} type - Node type (honest, attacker, intermittent)
     */
    constructor(id, joinDay = 0, type = 'honest') {
        this.id = id;
        this.joinDay = joinDay;
        this.type = type;

        // State tracking
        this.isOnline = true;
        this.participationDays = 0;
        this.offlineDays = 0;
        this.effectivenessWhenWentOffline = 0;
        this.totalRewards = 0;

        // History for analysis
        this.history = [];
    }

    /**
     * Update node state for a new day
     * @param {number} currentDay - Current simulation day
     * @returns {number} Current effectiveness
     */
    tick(currentDay) {
        if (currentDay < this.joinDay) {
            return 0;
        }

        let effectiveness;

        if (this.isOnline) {
            this.participationDays += 1;
            this.offlineDays = 0;
            effectiveness = calculateRamp(this.participationDays);
        } else {
            this.offlineDays += 1;
            effectiveness = calculateDecay(this.effectivenessWhenWentOffline, this.offlineDays);

            // If effectively reset, clear participation
            if (effectiveness === 0) {
                this.participationDays = 0;
            }
        }

        this.history.push({
            day: currentDay,
            effectiveness,
            isOnline: this.isOnline,
            participationDays: this.participationDays,
        });

        return effectiveness;
    }

    /**
     * Take node offline
     */
    goOffline() {
        if (this.isOnline) {
            this.effectivenessWhenWentOffline = calculateRamp(this.participationDays);
            this.isOnline = false;
        }
    }

    /**
     * Bring node back online
     */
    goOnline() {
        if (!this.isOnline) {
            // Calculate current effectiveness after decay
            const currentEff = calculateDecay(this.effectivenessWhenWentOffline, this.offlineDays);

            // Convert current effectiveness back to equivalent participation days
            // Eff(t) = 1 - exp(-t/R) => t = -R * ln(1 - Eff)
            if (currentEff > 0) {
                const R = 40; // From config
                this.participationDays = -R * Math.log(1 - currentEff);
            } else {
                this.participationDays = 0;
            }

            this.isOnline = true;
            this.offlineDays = 0;
        }
    }

    /**
     * Add rewards to node
     * @param {number} amount - Reward amount
     */
    addReward(amount) {
        this.totalRewards += amount;
    }

    /**
     * Get current effectiveness
     * @returns {number}
     */
    getEffectiveness() {
        if (this.history.length === 0) {
            return 0;
        }
        return this.history[this.history.length - 1].effectiveness;
    }

    /**
     * Get node summary
     * @returns {Object}
     */
    getSummary() {
        return {
            id: this.id,
            type: this.type,
            joinDay: this.joinDay,
            participationDays: this.participationDays,
            isOnline: this.isOnline,
            currentEffectiveness: this.getEffectiveness(),
            totalRewards: this.totalRewards,
        };
    }
}

/**
 * Factory for creating different node types
 */
export class NodeFactory {
    static nodeCounter = 0;

    /**
     * Create an honest node that stays online
     */
    static createHonest(joinDay = 0) {
        return new Node(`honest-${++this.nodeCounter}`, joinDay, 'honest');
    }

    /**
     * Create an intermittent node with random availability
     * @param {number} joinDay - Day to join
     * @param {number} uptimePercent - Percentage of time online (0-100)
     */
    static createIntermittent(joinDay = 0, uptimePercent = 80) {
        const node = new Node(`intermittent-${++this.nodeCounter}`, joinDay, 'intermittent');
        node.uptimePercent = uptimePercent;
        return node;
    }

    /**
     * Create an attacker node
     */
    static createAttacker(joinDay = 0) {
        return new Node(`attacker-${++this.nodeCounter}`, joinDay, 'attacker');
    }

    /**
     * Reset counter
     */
    static reset() {
        this.nodeCounter = 0;
    }
}

export default {
    Node,
    NodeFactory,
};
